source/controllers/include/controllers/Controller.hpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * @author Baptiste Busch
+ * @date 2021/02
+ *
+ */
+
+#pragma once
+
+#include "controllers/exceptions/NotImplementedException.hpp"
+#include "state_representation/Parameters/ParameterInterface.hpp"
+#include <list>
+#include <memory>
+
+namespace controllers {
+/**
+ * @class Controller
+ * @brief Abstract class to define a controller either in joint or cartesian spaces
+ * @tparam SIn the input state type of the controller
+ * @tparam SOut the output command type of the controller
+ */
+template <class SIn, class SOut>
+class Controller {
+public:
+  /**
+   * @brief Empty constructor
+   */
+  explicit Controller();
+
+  /**
+   * @brief Compute the command based on the input state
+   * To be redefined based on the actual controller implementation
+   * @param state the input state of the system. This function accept any number of extra arguments.
+   * @return the output command at the input state
+   */
+  virtual SOut compute_command(const SIn& state, ...) const;
```

Comment:
I wonder if a non-const version of compute_command would ever be needed, for a controller that has some history dependence (such as an integrating error term or something like that)

A derived class would not be able to override this with a non-const definition, since the signature is different, and so you also can't access that method of a derived object through a base pointer.

I guess for that you could add an `update()` method too, but that may complicate the pattern when derived classes have unique public functions. Some libraries use a pattern of `c_`, `nc_` prefix or suffix to denote const, non-const, and other return type variants for functions. Maybe there can be a `compute_command_nc` for that special case.

But anyway, it's a bit of a specific point and we don't need it just yet. Just something to think about.

Reply:
Ah, I thought of a very pertinent example:
the "passive" part of the passive DS controller is indeed non-const. It needs a time delta, and accordingly adapts its energy tank based on state.

Reply:
Yes totally agree with that. I actually had to remove the const for PassiveDS implementation. I guess we can probably make the two versions of the functions. Which also explain why I am not going with pure virtual in the `Controller` base class as I assumed that some derived class might not need to redefine all of the variations of the `compute_command` function.